### PR TITLE
Update Notifier

### PR DESCRIPTION
**Pull request by samanehsan**
### Purpose
Improve manually triggering analysis of HCA data to reduce the chance of running the wrong pipeline/data/etc.

### Changes

Querying for data to analyze:
- Update subscription queries in secondary-analysis (would be better to get these from Lira...)
- Generate a subscription query for the specified pipeline that matches on `files.project_json.provenance.document_id` to get all of the bundles for a project
- Check if the results contain multiple bundle versions and only return the newest ones

Sending notifications:
- Remove the "es_query" parameter from the notifier script (this is redundant and depends on the pipeline being run)
- Set the default vault path in the notifier script based on the specified environment

### Review Instructions
Any feedback on implementation/possible improvements are appreciated!

Note: If a project has multiple assays (e.g. SS2 and 10x) the script will need to be run twice

### CODE REVIEW

File: tests/scale_test/query_bundles_by_project_id.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+def format_query(project_uuid, workflow_name):
+    es_query_path = utils.SUBSCRIPTION_QUERIES.get(workflow_name)
+    if not es_query_path:
+        raise ValueError(f'No subscription query available for {workflow_name}.')
+    query = utils.load_es_query(es_query_path)
+    project_id_query = {'match': {'files.project_json.provenance.document_id': project_uuid}}
+    must_match = query['query']['bool']['must']
+    must_match.append(project_id_query)
+    query['query']['bool']['must'] = must_match
```
Comment (ApChagi) [inline tests/scale_test/query_bundles_by_project_id.py]: Can this just be `query['query']['bool']['must'].append(project_id_query)`?

File: tests/scale_test/query_bundles_by_project_id.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+def get_bundle_datetime(bundle_version):
+    return datetime.strptime(bundle_version, '%Y-%m-%dT%H%M%S.%fZ')
+
+def get_latest_bundle_versions(bundle_list):
+    bundle_ids = [b['bundle_uuid'] for b in bundle_list]
+    if len(bundle_ids) == len(set(bundle_ids)):
+        return bundle_list
+    else:
+        bundle_map = {}
```
Comment (ApChagi) [inline tests/scale_test/query_bundles_by_project_id.py]: I might just have too much clojure/fp on the brain, but I might do something like this here: 
```suggestion
from itertools import groupby
from functools import reduce

        ...
        keyfn = lambda x : x['bundle_uuid']
        sorted_bundles = sorted(bundle_list, key=keyfn)
        grouped_bundles = groupby(sorted_bundles, keyfn)
        return [reduce(choose_more_recent_bundle, g) for k, g in grouped_bundles]
```
where `choose_more_recent_bundle` would replace `get_bundle_datetime`

File: tests/scale_test/utils/utils.py [hunk 3]
```diff
@@ -349,6 +365,50 @@ def prepare_auth(lira_auth_dict):
...
+        dss_url (str): URL for the HCA Data Storage Service.
+        output_format (str): Format of the query results, either "summary" for a list of UUIDs or "raw" to include'
+            the bundle JSON metadata.
+        replica (str): The cloud replica to search in, either "gcp" or "aws".
+
+    Returns:
+        list: List of dicts in the format { bundle_uuid: <uuid>, bundle_version: <version> }
+
+    """
+    search_url = '{}/v1/search?output_format={}&replica={}&per_page=500'.format(
```
Comment (ApChagi) [inline tests/scale_test/utils/utils.py]: nit: could we use f'' here and below?

File: tests/scale_test/subscription_queries/smartseq2_query.json [hunk 4]
```diff
@@ -0,0 +1,785 @@
+{
```
Comment (pshapiro4broad) [inline tests/scale_test/subscription_queries/smartseq2_query.json]: These files are getting pretty large. Is there a to programatically generate this JSON that would take less lines of code?

File: tests/scale_test/utils/utils.py [hunk 5]
```diff
@@ -80,15 +87,18 @@ def prepare_notification(
...
+        es_query_path (str): The path to the ES query json file which is used for making subscription in BlueBox.
         label (str): A label to be added to the notification, which will then be added to the workflow started by Lira.
         transaction_id (str): A valid transaction id.
 
     Returns:
         notification (dict): A dict of valid notification content.
 
     """
+    if not es_query_path:
+        es_query_path = SUBSCRIPTION_QUERIES[workflow_name]
```
Comment (pshapiro4broad) [inline tests/scale_test/utils/utils.py]: If the workflow isn't in the list, and `es_query_path` isn't provided, should this raise an error?

File: tests/scale_test/query_bundles_by_project_id.py (hunk 2 above)
Comment (samanehsan) [reply tests/scale_test/query_bundles_by_project_id.py]: Thanks for the suggestion! It's much more elegant 😃

File: tests/scale_test/subscription_queries/smartseq2_query.json (hunk 4 above)
Comment (samanehsan) [reply tests/scale_test/subscription_queries/smartseq2_query.json]: Yes, there's a script in Lira that generates this!

Comment (samanehsan) [reply tests/scale_test/subscription_queries/smartseq2_query.json]: Unfortunately the queries are duplicated here...

File: tests/scale_test/utils/utils.py (hunk 5 above)
Comment (samanehsan) [reply tests/scale_test/utils/utils.py]: Good point!

File: tests/scale_test/analyze_project.sh [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    LIRA_URL="https://pipelines.data.humancellatlas.org"
+else
+    LIRA_URL="https://pipelines.${ENV}.data.humancellatlas.org"
+fi
+
+BUNDLE_LIST_FILE="${PROJECT_UUID}_bundles.json"
+LABELS="{'project_uuid': '${PROJECT_UUID}', 'project_shortname': '${PROJECT_SHORTNAME}'}"
+
+# Query for ${PROJECT_UUID} primary bundles in the ${ENV} data store using the subscription query for ${WORKFLOW_NAME}
+python query_bundles_by_project_id.py ${PROJECT_UUID} ${WORKFLOW_NAME} ${ENV} --output_file_path ${BUNDLE_LIST_FILE}
```
[the remark this answers is not in the document]
Comment (samanehsan) [reply tests/scale_test/analyze_project.sh]: That didn't seem to work for some reason 🤔

**Pull request merged**